Separated CircuitBreaker class into smaller independent state classes

kyewei opened the pull request:
Note: actual library changes is ~ 225 lines, the rest are new tests.

This PR contains changes to abstract out the AtomicInteger, AtomicSlidingWindow, and AtomicEnum classes with a shared SharedMemoryObject parent class. No new C code was added in this PR. 

The older, more massive PR was here: https://github.com/Shopify/semian/pull/54
See that for the general idea of how this is being designed towards.
@Sirupsen

fw42 commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,40 @@
+module Semian
+  module AtomicEnumSharedImplementation
```
conventional, this module should be called "AtomicEnum" or the filename should be "atomic_enum_shared_implementation.rb".

kyewei replied:
I would have thought that since the main detail of this file is implementing the `AtomicEnum` classes, whereas `AtomicEnumSharedImplementation` is just to facilitate sharing code, it should have been called atomic_enum. It's similar to `sliding_window_rb` in that it implements a class by that name. But alright.

fw42 commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,40 @@
+    end
+
+    def value
+      @num_to_sym.fetch super # May raise KeyError if num is not in list (invalid enum)
```
I don't think the comment adds much value... also parentheses please

fw42 commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,40 @@
+    end
+
+    def value=(sym)
+      super (@sym_to_num.fetch(sym)) # May raise KeyError if sym is not in list (invalid enum)
```
again, comment is unnecessary, imho.

fw42 commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,21 @@
+
+    def increase_by(val)
+      @value += val
+      self
```
I would expect this method to return the new value of `@value`, not `self`

fw42 commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,21 @@
+    end
+
+    def increase_by(val)
+      @value += val
```
hmm feels like this should call `self.value += val`, no? otherwise it's not using the `value=` method that you defined in the module..

fw42 commented on lib/semian/circuit_breaker.rb:
```diff
@@ -1,13 +1,37 @@
+      # (2) otherwise, keep the data
+    end
+
+    #Replacement for attr_reader :state
```
shouldn't this comment be on the `def state` method? ... imho comments like this are useless.. it's outdated already before you even shipped it :)

byroot commented:
This PR is ridded with basic syntax violations. I don't feel like nitpicking each of those, but I think we should do a quick rubocop integration.

fw42 commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,21 @@
+      @value=0
+    end
+
+    def increase_by(val)
```
can't you `def +(val)` this?

kyewei replied:
Maybe a different method name would make it clearer? The purpose is to increase the stored value by a set number. When the C implementation is added, keeping the entire mutating in one function is how an operation like `atomicinteger.value += 5` can maintain atomicity.

fw42 commented on lib/semian/shared_memory_object.rb:
```diff
@@ -0,0 +1,50 @@
+      _destroy if respond_to?(:_destroy) && @using_shared_memory
+    end
+
+    # Functions dealing with acquiring memory
```
I don't think we need a "function dealing with acquiring memory" comment on a function that's called "acquire_memory"

Sirupsen commented:
@byroot will that re-run here automatically?

byroot commented:
@Sirupsen you mean rubocop? After a rebase yes.

byroot commented on lib/semian/circuit_breaker.rb:
```diff
@@ -1,13 +1,35 @@
-    def initialize(exceptions:, success_threshold:, error_threshold:, error_timeout:)
+  class CircuitBreaker #:nodoc:
+    def initialize(name, exceptions:, success_threshold:, error_threshold:, error_timeout:, permissions: 0660)
+      @name = "#{name}"
```
Why not just `@name = name`? If you want to cast it to String, `name.to_s` will do.

byroot commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,39 @@
+  class SysVAtomicEnum < SysVAtomicInteger #:nodoc:
+    undef :increase_by
+    include AtomicEnumSharedImplementation
+  end
```
THis is weird. Why not have an `AbstractAtomicEnum`, and then `SysVAtomicEnum` and `AtomicEnum` inherits from it.

kyewei replied:
Can't do that since C methods will hook onto SysVAtomicInteger methods, so the SysV classes need to inherit from upper ones.

byroot commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,20 @@
+
+    def increase_by(val)
+      self.value += val
+    end
```
Beside tests I don't see why we would increment by anything else than 1. So I would at least provide a shorthand that is `increment`

kyewei replied:
You mean add another method that does the following? OK.
```
def increment
  increase_by(1)
end
```

byroot replied:
Yes, perfect.

byroot replied:
Also `increase_by` -> `increment_by` IMHO

Sirupsen commented on lib/semian.rb:
```diff
@@ -154,7 +157,12 @@ def resources
+require 'semian/atomic_integer'
+require 'semian/atomic_enum'
+Semian.extension_loaded = Semian.sysv_semaphores_supported? && Semian.semaphores_enabled?
+if Semian.extension_loaded
```
This basically means we now have three different ways to check for SysV: `Semian.sysv_semaphores_supported?`, `Semian.semaphores_enabled?` and now `Semian.extension_loaded`. IMO calling `Semian.semaphores_enabled?` should be enough here, but it will have to check that they're supported:

```ruby
  def semaphores_enabled?
    !ENV['SEMIAN_SEMAPHORES_DISABLED'] && Semian.sysv_semaphore_supported?
  end
```

Sirupsen commented on lib/semian/shared_memory_object.rb:
```diff
@@ -0,0 +1,48 @@
+module Semian
+  class SharedMemoryObject #:nodoc:
```
Why is all this SysV stuff in this PR? Why not start with just the Ruby part?

Sirupsen commented on lib/semian/circuit_breaker.rb:
```diff
@@ -1,13 +1,35 @@
+      @successes = ::Semian::SysVAtomicInteger.new(
+        "#{name}_atomic_integer",
+        permissions)
+      @state = ::Semian::SysVAtomicEnum.new(
```
Why is this SysV stuff in this PR?

kyewei replied:
This is the extent of all the ruby code that needs to be introduced, and it's small.

Sirupsen replied:
I think you need to split this PR in two yet again, and let the SysV stuff be reviewed independently.

Sirupsen commented:
Why is all this SysV stuff in this PR? Couldn't this PR just be using the `AtomicInteger`, `AtomicEnum` and `SlidingWindow` primitives from Ruby?

kyewei commented:
@Sirupsen it now only includes the primitives. It (tests, library, etc) has the effect of being less coupled now as well.

Sirupsen commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,37 @@
+module Semian
+  class AtomicEnum < AtomicInteger #:nodoc:
+    undef :increment_by
+    undef :increment
```
Why `undef` instead of overriding them with an exception? Does it actually matter that much even if they're defined? It'd work and would be equivalent to iterating on the enum, which is not a crazy thing to do.

kyewei replied:
I suppose to do that I'll have to add a `num%size` in to keep it within range.

Sirupsen replied:
Yeah. On the other hand, it wouldn't actually be used anywhere. I'm somewhat indifferent.

Sirupsen commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,37 @@
+    undef :increment
+
+    module AtomicEnumSharedImplementation
+      def initialize(name, permissions, symbol_list)
```
Similar comment on `name` and `permissions`.

kyewei replied:
What's the comment? I can't find it.

Sirupsen replied:
https://github.com/Shopify/semian/pull/56/files#r43757316

Sirupsen commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,37 @@
+    module AtomicEnumSharedImplementation
+      def initialize(name, permissions, symbol_list)
+        super(name, permissions)
+        initialize_lookup(symbol_list)
```
You could replace this with:

```
@num_to_sym = Hash[symbol_list.each_with_index.to_a].invert
```

Due to the well-named variable I think it's understandable.

kyewei replied:
So something like
```ruby
@sym_to_num = Hash[symbol_list.each_with_index.to_a]
@num_to_sym = @sym_to_num.invert
```
? I'd still like to keep it in a separate function though, since the SysV version will definitely reuse it

Sirupsen replied:
Yup, that's fine with me. I think you can give it a better name too.

Sirupsen commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,37 @@
+      end
+
+      def value=(sym)
+        super(@sym_to_num.fetch(sym))
```
:+1: on `#fetch` :-)

It would make sense to `alias_method` this to `#transition` or something similar, which would make the Circuit Breaker code easier to read. Yes / no?

Sirupsen commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,37 @@
+      end
+    end
+
+    include AtomicEnumSharedImplementation
```
I do not really understand why this has to be a module? The SysV implementation can just inherit from this, no?

kyewei replied:
There was a diamond problem in the fact that if we let SysVAtomicInteger inherit from AtomicInteger, and hook the methods in SysVAtomicInteger, to maintain both the Enum functionality and the SysV functionality in SysVAtomicEnum you'd have to get SysVAtomicEnum to einherit from SysVAtomicInteger and mixin the enum functionality... or hook twice.

Sirupsen replied:
Ah, I see what you mean.

kyewei replied:
I am thinking now that it may be possible to just use `extend Forwardable` and delegate. I usually strayed away from this because in most cases since there is delegation happening, lets say to an instance variable, the instance variable also needs to access content of the wrapper class, so there is back and forth communication, which is not good. However, in this case I think it works okay since the AtomicInteger instance doesn't need to know about the enum abstraction.

It'll just have a huge list of delegators in the end, something like this:
```
def_delegators :@integer, :semid, :shmid, :execute_atomically, :shared?,
                   :destroy, :acquire_memory_object, :bind_init_fn
```

Sirupsen commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,28 @@
+module Semian
+  class AtomicInteger #:nodoc:
```
Once we're on the SysV implementation we'll want to create a Thread Safe variant of this as well.

Sirupsen commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,28 @@
+  class AtomicInteger #:nodoc:
+    attr_accessor :value
+
+    def initialize(_name, _permissions)
```
Why the underscored arguments here? These don't matter for `AtomicInteger`. They only matter for the SysV adapter which will override the initializer regardless.

kyewei replied:
Are you suggesting just making it a 0 parameter initializer? I'd like to keep the interface the same though.

Sirupsen replied:
Yes. I don't have a good idea of what this will look like, but the in-memory Ruby adapter can't take permissions just because another adapter happens to need it.

kyewei replied:
Perhaps with keyword arguments? For example:
```ruby
class AtomicEnum
  def initialize(symbol_list)
  ...

class SysVAtomicEnum < AtomicEnum
  def initialize(symbol_list, name: , permissions: )
  ...

```
This would work. Would there be any compatibility issues?

Sirupsen commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,28 @@
+
+    def increment
+      increment_by(1)
+    end
```
It'd be simpler (and more common in Ruby) to just do:

```ruby
def increment(val = 1)
  value += val
end
```

Sirupsen commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,28 @@
+
+    def execute_atomically
+      yield if block_given?
+    end
```
I'm not a huge fan of this method name, but I see what you're trying to do. How about just `#transaction`?

kyewei replied:
Are you referencing something like this? http://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html
I've never personally seen that use but if it's idiomatic I'll put that in.

Sirupsen replied:
Yeah, it's common in AR, and e.g. `PStore` in the std lib uses that name as well.

fw42 replied:
I don't understand this method or this name... I don't see anything begin atomic.. Can you explain?

kyewei replied:
Look at this first: https://github.com/Shopify/semian/pull/54#discussion-diff-42760663

I don't think we've decided yet what to do with it in this case of AtomicInteger, its part of `SharedMemoryObject#execute_atomically`

kyewei replied:
The reason I found this "separating the big PR into small PR's" tedious is that some code is placed out of context and I have to re-explain every implementation reasoning that would otherwise make perfect sense as a whole, or redirect people to PR https://github.com/Shopify/semian/pull/54 :(

fw42 replied:
I see. I didn't even know there was another PR.

Sirupsen commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,28 @@
+
+    def shared?
+      false
+    end
```
Why is this important? It's not actually used anywhere (yet?).

kyewei replied:
If we're going to add a shared version, then we need to be able to distinguish between for example `AtomicInteger`, `SysVAtomicInteger`, and the fallback `SysVAtomicInteger` on a semaphore-less system.

Sirupsen replied:
Why does the distinction matter?

kyewei replied:
I'd think that users would want to know if their structures are shared or not, and also that the data may be changed from a different worker. It also makes testing shared components easier since you can test the expected behaviour when you change another structure with the same name.

Sirupsen replied:
Why would the user care by calling a method? They'd find out which adapter is used, but they don't need to pragmatically know.

I don't understand why it's easier to test. Each adapter should be tested independently.

kyewei replied:
The user should care because they might want to handle the two cases that can occur (shared, not shared) separately. 

Easier to test and less repeated code through something like this:
```ruby
def test_shared_attribute
  run_test_with_sliding_window_classes do |klass, instance|
    return unless instance.shared?
    # test contents
  end
end
```
There is no point in writing the same code x times, once for each type of class and once for a shared vs not-shared instance of a class. 

How else can you distinguish between a shared-memory version of a class and a not-shared-memory version? There is no other method that can distinguish this without explicitly testing to see if it is. When the `SharedMemoryObject` superclass comes in, `shared?`, `execute_atomically`, and `destroy` will be moved there. 

I'm pretty adamant about this being included since many later tests for testing the shared component depend on this as well and otherwise there is no way to distinguish them making writing tests more difficult than it should be. If the issue is that it doesn't make sense to include it at this iteration, but that it will be included when I add in the shared memory class, then that's OK.

Sirupsen replied:
I'm not a big fan of this pattern for testing adapters. At least the classes to test should be hardcoded. @byroot can you weight in here for a third opinion? Do you think having `#shared` and iterating over all the adapter classes and sharing the tests is OK?

byroot replied:
Sharing tests are a good idea. A single test testing them all is not.

What I would recommend is simply using a module:

```ruby
class SysVAtomicIntegerTest < Minitest::TestCase
  include SharedAtomicIntegerTestCases

  # Add tests specific to SysVAtomicInteger here.
end
```

kyewei replied:
Here is the type of code I have ready for the iteration after this one for this type of purpose:
```ruby
def test_memory_is_shared
  run_test_with_atomic_integer_classes do |klass|
    next unless @integer.shared?
    integer_2 = klass.new('TestAtomicInteger', 0660)
    integer_2.value = 100
    assert_equal 100, @integer.value
    @integer.value = 200
    assert_equal 200, integer_2.value
    @integer.value = 0
    assert_equal 0, integer_2.value
  end
end

def test_memory_not_reset_when_at_least_one_worker_using_it
  run_test_with_atomic_integer_classes do |klass|
    next unless @integer.shared?

    @integer.value = 109
    integer_2 = klass.new('TestAtomicInteger', 0660)
    assert_equal @integer.value, integer_2.value
    pid = fork do
      integer_3 = klass.new('TestAtomicInteger', 0660)
      assert_equal 109, integer_3.value
      sleep
    end
    sleep 1
    Process.kill("KILL", pid)
    Process.waitall
    fork do
      integer_3 = klass.new('TestAtomicInteger', 0660)
      assert_equal 109, integer_3.value
    end
    Process.waitall
  end
end

private

def atomic_integer_classes
  @classes ||= retrieve_descendants(::Semian::AtomicInteger)
end

def run_test_with_atomic_integer_classes(klasses = atomic_integer_classes)
  klasses.each do |klass|
    begin
      @integer = klass.new('TestAtomicInteger', 0660)
      @integer.value = 0
      yield(klass)
    ensure
      @integer.destroy
    end
  end
end


```

With this you can test the structures without regard to whether it is a shared `SysVAtomicInteger`, a non-shared (disabled or OS not supporting) `SysVAtomicInteger`, or a regular `AtomicInteger`. Furthermore, you can even make a subclass like this to emulate shareness for testing correctness purposes (since on non-semaphore systems, you want to have at least one set of tests run) as long as `#shared?` is implemented and is overriden to be true:
```ruby
class FakeSysVAtomicInteger < Semian::AtomicInteger
  class << self
    attr_accessor :resources
  end
  self.resources = {}
  attr_accessor :name
  def self.new(name, _permissions)
    obj = resources[name] ||= super
    obj.name = name
    obj
  end

  def destroy
    self.class.resources.delete(@name)
    super
  end

  def shared?
    true
  end
end
```
Without something like `#shared?` it is hard to do a thorough test that captures all possible conditions without repeating code excessively.

kyewei replied:
^See above for more.

So @byroot, how would you handle testing `SysVAtomicInteger` on a system that doesn't support semaphores? You would expect `SysVAtomicIntegerTest` to have assertions that things like
```
integer_1 = Semian::SysVAtomicInteger.new("example", 0660)
integer_2 = Semian::SysVAtomicInteger.new("example", 0660)
integer_2.value = 2
assert_equal(integer_2.value, integer_1.value)
```
That last line would definitely fail since you're running a fallback code while expecting it not to have fallback.

byroot replied:
> how would you handle testing SysVAtomicInteger on a system that doesn't support semaphores?

You don't test it. It's what we always did. 

> next unless shared? then it would be fine.

If that is what shared is for, then it's bad naming. 
I encourage you to take a look at the `thread_safe` gem, on how they handle code radically different based on the platform: https://github.com/ruby-concurrency/thread_safe

kyewei replied:
Yes, essentially I want something that lets me differentiate between a structure that has shared memory (and therefore needs a different set of methods for `destroy` (since we now need to free the semaphore, etc), `execute_atomically` (which locks, calls yield, and unlocks), but because we now have a fallback implementation as well, this functionality can't just be a complete override, so we need some `bool` of some sort to determine whether we are actually using shared memory. Thats what the `#shared?` was for. 

I can concede the iterating over all the classes for testing part, but that `bool` has to be there for things like `destroy` and `execute_atomically`. `#shared?` can be private if that seems like a better idea.

kyewei replied:
So what's the consensus for `#shared?`? @byroot @Sirupsen. The most important reason is outlined above:
> Yes, essentially I want something that lets me differentiate between a structure that has shared memory (and therefore needs a different set of methods for `destroy` (since we now need to free the semaphore, etc), `execute_atomically` (which locks, calls yield, and unlocks). However, because we now have a fallback implementation as well, this functionality can't just be a complete override, so we need some `bool` of some sort to determine whether we are actually using shared memory and branch off. Thats what the `#shared?` was for.

If `#shared?` isn't included (it could be `private` for all I care) then I need a different way of implementing falbacks.

byroot replied:
If it's private then I don't mind.

Sirupsen commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,28 @@
+    end
+
+    def destroy
+    end
```
Should this set the value back to 0?

Sirupsen commented on lib/semian/circuit_breaker.rb:
```diff
@@ -1,13 +1,35 @@
-    def initialize(exceptions:, success_threshold:, error_threshold:, error_timeout:)
+  class CircuitBreaker #:nodoc:
+    def initialize(name, exceptions:, success_threshold:, error_threshold:, error_timeout:, permissions: 0660)
+      @name = name.to_s
```
Can you explain why it needs a name now? Is this purely so you have something to hash on for the SysV implementation? Wouldn't it be better to pass a reference to the resource which has a name?

kyewei replied:
It's so that the individual structures have a name. If you want to invert it, my argument against that would be that you'd need to have this code:
```
::Semian::AtomicEnum.new(
        name,
        permissions,
        [:closed, :half_open, :open])
```
outside the circuit breaker class and would be that wouldn't be great.

Sirupsen commented on lib/semian/circuit_breaker.rb:
```diff
@@ -1,13 +1,35 @@
       @exceptions = exceptions
-      reset
+
+      @sliding_window = ::Semian::SlidingWindow.new(
```
Why leak the abstraction of the data structure when you could retain the previous name of `@errors`?

Sirupsen commented on lib/semian/circuit_breaker.rb:
```diff
@@ -1,13 +1,35 @@
+        permissions)
+      @state = ::Semian::AtomicEnum.new(
+        "#{name}_atomic_enum",
+        permissions,
```
Permissions don't make sense here.

kyewei replied:
What do you mean?

Sirupsen replied:
Permissions don't make sense for a Ruby implementation. This is adapter specific.

Sirupsen commented on lib/semian/circuit_breaker.rb:
```diff
@@ -1,13 +1,35 @@
+    end
+
+    def shared?
+      @sliding_window.shared? && @successes.shared? && @state.shared?
```
Similarly, why does this need to be shared?

Sirupsen commented on lib/semian/circuit_breaker.rb:
```diff
@@ -1,13 +1,35 @@
+    end
+
+    def state
+      @state.value
```
You never use this.

Sirupsen commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+      @window = []
+    end
+
+    def execute_atomically
```
Similar concern about naming.

Sirupsen commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+      self
+    end
+
+    def unshift(time_ms)
```
Calling this `#prepend` is more true to the data structure IMO. But why do we even need this method? If this is purely for testing (which it seems to be) it should be a helper instead of a public method on the data structure.

Sirupsen commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+    def push(time_ms)
+      @window.shift while @window.size >= @max_size
+      @window << time_ms
+      self
```
Why return `self`?

kyewei replied:
Maintain the abstraction that the wrapper is responsible for handling appending, even if it is delegating it to an actual array. In `sliding_window << 1 << 2`, if there was no self, then after the first push it would leak the array.

Sirupsen commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+
+    def shared?
+      false
+    end
```
Again, why is this important? Why does this abstraction have to leak?

Sirupsen commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+
+    def destroy
+    end
```
Call `#reset`?

Sirupsen commented on lib/semian/unprotected_resource.rb:
```diff
@@ -36,5 +36,12 @@ def request_allowed?
+
+    def circuit_breaker_shared?
+      false
+    end
```
Why does this matter?

kyewei replied:
If... you're referring to `circuit_breaker_shared?`, then I'm going to remove it along with making `#shared?` private. If it was the `mark_success`, there was a test somewhere that checked the `#public_instance_methods` of both types of the `Resource` wrappers to make sure they were the same.

Sirupsen commented on test/atomic_enum_test.rb:
```diff
@@ -0,0 +1,35 @@
+    @classes ||= retrieve_descendants(::Semian::AtomicEnum)
+  end
+
+  def run_test_with_atomic_enum_classes(klasses = atomic_enum_classes)
```
This is a little crazy :-) I think we should just write tests for each one. They may have different tests too, e.g. guarantee they're atomic, shared, and so on.

Sirupsen commented on test/atomic_enum_test.rb:
```diff
@@ -0,0 +1,35 @@
+require 'test_helper'
+
+class TestAtomicEnum < MiniTest::Unit::TestCase
```
When writing unit tests, I like to browse the class I'm testing and look at each line and think: If I delete this line of code, will a test fail? If I change this greater than to a less than, would a test fail? What if it's a less than or equal?

I bet you can find a lot of lines for which that's not true and increase test coverage.

fw42 replied:
This class should be called `AtomicEnumTest` for consistent naming (the file is called `atomic_enum_test.rb`)

kyewei replied:
Actually, if you look at the naming of all the test files in this folder, it's written like this.

fw42 replied:
hrm ok if all the other tests are already doing it wrong then leave it like this... :-(

kyewei replied:
I kinda dislike doing things like that prematurely since I assume they're there for a reason. If I do, someone always dislikes the change (and is usually very vocal about it). But OK, I'll do that after.

Sirupsen commented on test/atomic_integer_test.rb:
```diff
@@ -0,0 +1,57 @@
+    @classes ||= retrieve_descendants(::Semian::AtomicInteger).select { |klass| /Integer/.match(klass.name) }
+  end
+
+  def run_test_with_atomic_integer_classes(klasses = atomic_integer_classes)
```
Similar note.

Sirupsen commented on test/atomic_integer_test.rb:
```diff
@@ -0,0 +1,57 @@
+    end
+  end
+
+  def test_execute_atomically_actually_is_atomic
```
Why all this when the implementation doesn't even have a mutex?

Sirupsen commented on test/atomic_integer_test.rb:
```diff
@@ -0,0 +1,57 @@
+require 'test_helper'
+
+class TestAtomicInteger < MiniTest::Unit::TestCase
+  def test_operations
```
When doing a lot of assertions in one test in a unit test, that's usually a smell you could have more well-named tests. Note that one of those may be exactly equal to this one, but has a good name on what you're trying to test and perhaps only one or two assertions.

Sirupsen commented on test/sliding_window_test.rb:
```diff
@@ -0,0 +1,63 @@
+    @classes ||= retrieve_descendants(::Semian::SlidingWindow)
+  end
+
+  def run_test_with_sliding_window_classes(klasses = sliding_window_classes)
```
Similar comment.

Sirupsen commented on test/sliding_window_test.rb:
```diff
@@ -0,0 +1,63 @@
+require 'test_helper'
+
+class TestSlidingWindow < MiniTest::Unit::TestCase
+  def test_sliding_window_functionality
```
Similar comments on more tests + smaller tests.

Sirupsen commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+    # like this: if @max_size = 4, current time is 10, @window =[5,7,9,10].
+    # Another push of (11) at 11 sec would make @window [7,9,10,11], shifting off 5.
+
+    def resize_to(size)
```
You only use this in tests. Shouldn't this be used instead of the very similar implementation in `push_time`?

kyewei replied:
This one changes the max_size as well. Although, the similar implementation in `push_time` has this line, which is what I think you're referring to:
```ruby
window.shift if window.size == max_size
```
I realized that line is redundant since `#<<` manages this already, so I'm removing that one line.

Sirupsen commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+      throw ArgumentError.new('size must be larger than 0') if size < 1
+      @max_size = size
+      @window.shift while @window.size > @max_size
+      self
```
Why return self?

kyewei replied:
I put it in so cascading could work `window.resize_to(3).pop(1)`. 
Would returning the new size be more logical? It'll prevent ^ from being possible though.

Sirupsen commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+    # Another push of (11) at 11 sec would make @window [7,9,10,11], shifting off 5.
+
+    def resize_to(size)
+      throw ArgumentError.new('size must be larger than 0') if size < 1
```
We don't use `throw` and `try` in Shopify Ruby, use `raise`.

kyewei commented:
This is probably good for another review. Things I changed: testing style, the initializer (now has keyword arguments for `name:`, `permissions:`, fixed `destroy`, made `shared?` private, added more tests.

fw42 commented on lib/semian.rb:
```diff
@@ -162,7 +167,7 @@ def resources
   end
 
-  unless Semian.semaphores_enabled?
+  if ENV['SEMIAN_SEMAPHORES_DISABLED']
```
why this change? seems more reasonable to use the method instead of check the env var again

kyewei replied:
This was done a few commits ago, when Simon suggested `#semaphores_enabled?` should include `&& Semian.sysv_semaphores_supported?`, which means the line here needs to be changed.

fw42 commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,39 @@
+    def_delegators :@integer, :execute_atomically, :transaction, :shared?, :destroy
+    private :shared?
+
+    def initialize(symbol_list, **_options)
```
I don't understand, why does this have an `options` parameter? You are not using those values anywhere, are you?

kyewei replied:
I wanted to keep the interface the same. Otherwise things like `name: 'blah'`, etc in the initializer will throw errors.

fw42 replied:
What interface? I don't understand.. Why would you give this thing a `name: 'bla'` parameter if the class is not doing anything with it?

kyewei replied:
When the SysV stuff come in, they need a name to hash on, and permissions to use for the `semget` functions, etc. I was talking about how it keeps the interface the same if the non SysV structures accept the same, but just don't do anything with it. But if you give keyword arguments to an initializer that doesn't accept keyword arguments it throws an exception.

fw42 commented on lib/semian/atomic_integer.rb:
```diff
@@ -0,0 +1,29 @@
+  class AtomicInteger #:nodoc:
+    attr_accessor :value
+
+    def initialize(**_options)
```
What's with the options again? Why does this exist if it's not used?

fw42 commented on lib/semian/circuit_breaker.rb:
```diff
@@ -1,13 +1,24 @@
+      # We do not need to #reset here since initializing is handled like this:
+      # (0) if data is not shared, then it's zeroed already
+      # (1) if no one is attached to the memory, zero it
+      # (2) otherwise, keep the data
```
This comment is gonna get outdated... also I don't understand it... "if no one is attached to the memory, zero it"? what does attached mean? and "who" would be attached to things? this is very confusing, imho

kyewei replied:
That... applies to the C portion. I got rid of it.

fw42 commented on lib/semian/circuit_breaker.rb:
```diff
@@ -43,70 +53,78 @@ def mark_failed(_error)
-      window.shift if window.size == max_size
-      window << time
+    def push_time(window, duration:, time: Time.now)
+      @errors.execute_atomically do # Store an integer amount of milliseconds since epoch
```
what is this comment referring to? to this line? or the next line? or this whole method? I don't think this adds much value

fw42 commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+    def_delegators :@window, :size, :pop, :shift, :first, :last
+    attr_reader :max_size
+
+    def initialize(max_size, **_)
```
`**_`? I don't understand

fw42 commented on lib/semian/sliding_window.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+
+    def clear
+      @window = []
```
could use `@window.clear` to save the unnecessary allocation..

fw42 commented on test/test_helper.rb:
```diff
@@ -6,6 +6,7 @@
 require 'timecop'
 require 'tempfile'
 require 'fileutils'
+require 'timeout'
```
is this used anywhere?

kyewei commented:
Any other issues?

kyewei commented:
I rebased + removed `#shared`, `#execute_atomically`. There should be no more code that doesn't make sense in context.

byroot commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,38 @@
+    end
+
+    def increment(val = 1)
+      @integer.value = (@integer.value + val) % @sym_to_num.size
```
How is this atomic? You read the value, do your computation in memory, and then assign it back.

Am I missing something?

I'm also not sure why you'd what to "increment" an Enum.

kyewei replied:
Simon mentioned iterating through the enums like this, I originally had `undef :increment`, but he said he didn't care whether the iterating should be added or not, so I could remove it.

kyewei replied:
Should I add a mutex-lock at this stage? https://github.com/Shopify/semian/pull/56/files#r43757308

byroot replied:
It's what I'm trying to say. Even if your `AtomicInteger` was actually atomic, the way to you use it is not.

byroot replied:
But maybe I'm missing something. Are all these classes just stubs?

kyewei replied:
Most are really small stubs, since when I put in the SharedMemoryObject stuff, Simon said that was already too big, so I had to rip that out, and so the mutex stuff, which I would put in there, is gone. Maybe a rename to just `Integer` than `AtomicInteger` ?

kyewei replied:
Here's what it should look like: 
https://github.com/Shopify/semian/blob/circuit-breaker-per-host/lib/semian/atomic_integer.rb

Notice the `surround_with_mutex`, ... but I threw that in `SharedMemoryObject`

byroot commented on lib/semian/atomic_enum.rb:
```diff
@@ -0,0 +1,38 @@
+      @sym_to_num = Hash[symbol_list.each_with_index.to_a]
+      @num_to_sym = @sym_to_num.invert
+    end
+  end
```
Do you really need that Enum to be generalized? I think the code would read easier if the enum would clearly limit itself to the 3 states we need.

kyewei replied:
I've been working towards this for a while now, the motivation was to separate/abstract out the classes into distinct structures. 

What part of the code is difficult to read? Maybe that is the issue here instead. Is it the one liner on line 34? See this: https://github.com/Shopify/semian/pull/56#discussion_r43757296
